File: tutorial/worked-examples/bac-replication-r/analysis/code/03_event_study.R
# 03_event_study.R - Event study specification

library(fixest)
library(dplyr)
library(tidyr)

# Load analysis data
analysis_data <- readRDS(file.path(build, "output", "analysis_data.rds"))

# Build formula with available controls
# Policy controls are always present (created in build script)
controls <- c("alr", "zero_tolerance", "primary_seatbelt", "secondary_seatbelt",
              "mlda21", "gdl", "speed_70", "aggravated_dui")
if ("unemployment" %in% names(analysis_data)) {
  controls <- c(controls, "unemployment")
}
if ("income" %in% names(analysis_data)) {
  controls <- c(controls, "income")
}

# Use fixest's bin argument to bin endpoint event times
# instead of manually creating event_time_binned with case_when
my_bin <- .("-5+" = ~x <= -5, "10+" = ~x >= 10)

# Event study: Hit-and-run
cat("  Running event study: ln_hr ~ event_time_dummies + FE\n")
if (length(controls) > 0) {
  hr_es_formula <- as.formula(paste(
    "ln_hr ~ i(event_time, ref = -1, bin = my_bin) +",
    paste(controls, collapse = " + "),
    "| state_fips + year"
  ))
} else {
  hr_es_formula <- ln_hr ~ i(event_time, ref = -1, bin = my_bin) | state_fips + year
}
hr_es_model <- feols(hr_es_formula, data = analysis_data, cluster = ~state_fips)

# Extract coefficients for hit-run using broom::tidy
hr_coefs <- broom::tidy(hr_es_model, conf.int = TRUE) %>%
  filter(grepl("event_time", term)) %>%
  mutate(
    event_time = as.numeric(stringr::str_extract(term, "-?\\d+")),
    pvalue = p.value
  ) %>%
  select(event_time, coefficient = estimate, std_error = std.error,
         pvalue, ci_lower = conf.low, ci_upper = conf.high) %>%
  arrange(event_time)

# Add reference period
hr_coefs <- bind_rows(
  hr_coefs,
  tibble(event_time = -1, coefficient = 0, std_error = 0, pvalue = NA,
         ci_lower = 0, ci_upper = 0)
) %>%
  arrange(event_time)

# Save hit-run coefficients
write.csv(hr_coefs, file.path(analysis, "output", "tables", "es_coefficients_hr.csv"),
          row.names = FALSE)

# Print hit-run results
cat("  Event study coefficients (Hit-Run):\n")
for (i in 1:nrow(hr_coefs)) {
  row <- hr_coefs[i, ]
  sig <- ifelse(!is.na(row$pvalue) && row$pvalue < 0.05, "*", "")
  cat(sprintf("    t=%+3d: %7.4f (%6.4f)%s\n",
              as.integer(row$event_time), row$coefficient, row$std_error, sig))
}

# Event study: Non-hit-and-run
cat("\n  Running event study: ln_nhr ~ event_time_dummies + FE\n")
if (length(controls) > 0) {
  nhr_es_formula <- as.formula(paste(
    "ln_nhr ~ i(event_time, ref = -1, bin = my_bin) +",
    paste(controls, collapse = " + "),
    "| state_fips + year"
  ))
} else {
  nhr_es_formula <- ln_nhr ~ i(event_time, ref = -1, bin = my_bin) | state_fips + year
}
nhr_es_model <- feols(nhr_es_formula, data = analysis_data, cluster = ~state_fips)

# Extract coefficients for non-hit-run
nhr_coefs <- broom::tidy(nhr_es_model, conf.int = TRUE) %>%
  filter(grepl("event_time", term)) %>%
  mutate(
    event_time = as.numeric(stringr::str_extract(term, "-?\\d+")),
    pvalue = p.value
  ) %>%
  select(event_time, coefficient = estimate, std_error = std.error,
         pvalue, ci_lower = conf.low, ci_upper = conf.high) %>%
  arrange(event_time)

# Add reference period
nhr_coefs <- bind_rows(
  nhr_coefs,
  tibble(event_time = -1, coefficient = 0, std_error = 0, pvalue = NA,
         ci_lower = 0, ci_upper = 0)
) %>%
  arrange(event_time)

# Save non-hit-run coefficients
write.csv(nhr_coefs, file.path(analysis, "output", "tables", "es_coefficients_nhr.csv"),
          row.names = FALSE)

# Store for figures script
ES_COEF_HR <- hr_coefs
ES_COEF_NHR <- nhr_coefs

cat("  Saved event study coefficients\n")

File: tutorial/worked-examples/bac-replication-python/analysis/code/03_event_study.py
# 03_event_study.py - Event study specification

import pandas as pd
import numpy as np
import pyfixest as pf

# Load analysis data
analysis_data = pd.read_parquet(BUILD / "output" / "analysis_data.parquet")

# Bin event time at endpoints (-5 to +10)
MIN_ET, MAX_ET = -5, 10
analysis_data['event_time_binned'] = analysis_data['event_time'].copy()
analysis_data.loc[analysis_data['event_time'] < MIN_ET, 'event_time_binned'] = MIN_ET
analysis_data.loc[analysis_data['event_time'] > MAX_ET, 'event_time_binned'] = MAX_ET

# Get unique event times (excluding -1 as reference)
event_times = sorted([et for et in analysis_data['event_time_binned'].dropna().unique() if et != -1])

# Create event time dummies with safe column names
for et in event_times:
    safe_name = f'et_m{abs(int(et))}' if et < 0 else f'et_p{int(et)}'
    analysis_data[safe_name] = (analysis_data['event_time_binned'] == et).astype(int)

# Build formula
et_cols = [f'et_m{abs(int(et))}' if et < 0 else f'et_p{int(et)}' for et in event_times]
# Policy controls are always present (created in build script)
controls = ['alr', 'zero_tolerance', 'primary_seatbelt', 'secondary_seatbelt',
            'mlda21', 'gdl', 'speed_70', 'aggravated_dui']
if 'unemployment' in analysis_data.columns:
    controls.append('unemployment')
if 'income' in analysis_data.columns:
    controls.append('income')
all_vars = et_cols + controls
formula_hr = f'ln_hr ~ {" + ".join(all_vars)} | state_fips + year'

print(f"  Running event study: ln_hr ~ event_time_dummies + FE")

# Fit model
hr_results = pf.feols(formula_hr, data=analysis_data, vcov={'CRV1': 'state_fips'})

# Extract coefficients
coefs = []
for et in event_times:
    safe_name = f'et_m{abs(int(et))}' if et < 0 else f'et_p{int(et)}'
    coefs.append({
        'event_time': et,
        'coefficient': hr_results.coef()[safe_name],
        'std_error': hr_results.se()[safe_name],
        'pvalue': hr_results.pvalue()[safe_name],
        'ci_lower': hr_results.coef()[safe_name] - 1.96 * hr_results.se()[safe_name],
        'ci_upper': hr_results.coef()[safe_name] + 1.96 * hr_results.se()[safe_name]
    })

# Add reference period
coefs.append({
    'event_time': -1,
    'coefficient': 0,
    'std_error': 0,
    'pvalue': np.nan,
    'ci_lower': 0,
    'ci_upper': 0
})

coef_df_hr = pd.DataFrame(coefs).sort_values('event_time')
coef_df_hr.to_csv(ANALYSIS / "output" / "tables" / "es_coefficients_hr.csv", index=False)

print("  Event study coefficients (Hit-Run):")
for _, row in coef_df_hr.iterrows():
    sig = '*' if (row['std_error'] > 0 and abs(row['coefficient'] / row['std_error']) > 1.96) else ''
    print(f"    t={int(row['event_time']):+3d}: {row['coefficient']:7.4f} ({row['std_error']:.4f}){sig}")

# Also run for non-hit-run
formula_nhr = f'ln_nhr ~ {" + ".join(all_vars)} | state_fips + year'
nhr_results = pf.feols(formula_nhr, data=analysis_data, vcov={'CRV1': 'state_fips'})

coefs_nhr = []
for et in event_times:
    safe_name = f'et_m{abs(int(et))}' if et < 0 else f'et_p{int(et)}'
    coefs_nhr.append({
        'event_time': et,
        'coefficient': nhr_results.coef()[safe_name],
        'std_error': nhr_results.se()[safe_name],
        'pvalue': nhr_results.pvalue()[safe_name],
        'ci_lower': nhr_results.coef()[safe_name] - 1.96 * nhr_results.se()[safe_name],
        'ci_upper': nhr_results.coef()[safe_name] + 1.96 * nhr_results.se()[safe_name]
    })

coefs_nhr.append({
    'event_time': -1,
    'coefficient': 0,
    'std_error': 0,
    'pvalue': np.nan,
    'ci_lower': 0,
    'ci_upper': 0
})

coef_df_nhr = pd.DataFrame(coefs_nhr).sort_values('event_time')
coef_df_nhr.to_csv(ANALYSIS / "output" / "tables" / "es_coefficients_nhr.csv", index=False)

# Store for figures script
ES_COEF_HR = coef_df_hr
ES_COEF_NHR = coef_df_nhr

print("  Saved event study coefficients")
